src/util/strencodings.h

```diff
@@ -365,4 +367,79 @@ std::string Capitalize(std::string str);
  */
 std::optional<uint64_t> ParseByteUnits(std::string_view str, ByteUnit default_multiplier);
 
+namespace util {
+/** consteval version of HexDigit() without the lookup table. */
+consteval uint8_t ConstevalHexDigit(const char c)
+{
+    if (c >= '0' && c <= '9') return c - '0';
+    if (c >= 'a' && c <= 'f') return c - 'a' + 0xa;
+
+    throw "Only lowercase hex digits are allowed, for consistency";
+}
+
+/**
+ * ""_hex is a compile-time user-defined literal returning a
+ * `std::array<std::byte>`, equivalent to ParseHex(). Variants provided:
+ *
+ * - ""_hex_v: Returns `std::vector<std::byte>`, useful for heap allocation or
+ *   variable-length serialization.
+ *
+ * - ""_hex_u8: Returns `std::array<uint8_t>`, for cases where `std::byte` is
+ *   incompatible.
+ *
+ * - ""_hex_v_u8: Returns `std::vector<uint8_t>`, combining heap allocation with
+ *   `uint8_t`.
+ *
+ * @warning It could be necessary to use vector instead of array variants when
+ *   serializing, or vice versa, because vectors are assumed to be variable-
+ *   length and serialized with a size prefix, while arrays are considered fixed
+ *   length and serialized with no prefix.
+ *
+ * @warning It may be preferable to use vector variants to save stack space when
+ *   declaring local variables if hex strings are large. Alternately variables
+ *   could be declared constexpr to avoid using stack space.
+ *
+ * @warning Avoid `uint8_t` variants when not necessary, as the codebase
+ *   migrates to use `std::byte` instead of `unsigned char` and `uint8_t`.
+ *
+ * @note One reason ""_hex uses `std::array` instead of `std::vector` like
+ *   ParseHex() does is because heap-based containers cannot cross the compile-
+ *   time/runtime barrier.
+ */
+inline namespace hex_literals {
+namespace detail {
+
+template <size_t N>
+struct Hex {
+    std::array<std::byte, N / 2> bytes{};
+    consteval Hex(const char (&hex_str)[N])
+        // 2 hex digits required per byte + implicit null terminator
+        requires(N % 2 == 1)
+    {
+        if (hex_str[N - 1]) throw "null terminator required";
+        for (std::size_t i = 0; i < bytes.size(); ++i) {
+            bytes[i] = static_cast<std::byte>(
+                (ConstevalHexDigit(hex_str[2 * i]) << 4) |
+                 ConstevalHexDigit(hex_str[2 * i + 1]));
+        }
+    }
+};
+
+} // namespace detail
+
+template <util::detail::Hex str>
+constexpr auto operator""_hex() { return str.bytes; }
+
+template <util::detail::Hex str>
+constexpr auto operator""_hex_u8() { return std::bit_cast<std::array<uint8_t, str.bytes.size()>>(str.bytes); }
+
+template <util::detail::Hex str>
+constexpr auto operator""_hex_v() { return std::vector<std::byte>{str.bytes.begin(), str.bytes.end()}; }
```

Comment:
Note: the `_hex_v` operator is currently unused, except for in unit tests on this operator:

<details>
<summary>git diff on dae42d474d</summary>

```diff
diff --git a/src/test/util_tests.cpp b/src/test/util_tests.cpp
index 46eca2f0d8..ab521f5628 100644
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -158,10 +158,6 @@ BOOST_AUTO_TEST_CASE(parse_hex)
     auto hex_literal_span{MakeUCharSpan(hex_literal_array)};
     BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_span.begin(), hex_literal_span.end(), expected.begin(), expected.end());
 
-    const std::vector<std::byte> hex_literal_vector{operator""_hex_v<util::detail::Hex(HEX_PARSE_INPUT)>()};
-    hex_literal_span = MakeUCharSpan(hex_literal_vector);
-    BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_span.begin(), hex_literal_span.end(), expected.begin(), expected.end());
-
     constexpr std::array<uint8_t, 65> hex_literal_array_uint8{operator""_hex_u8<util::detail::Hex(HEX_PARSE_INPUT)>()};
     BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_array_uint8.begin(), hex_literal_array_uint8.end(), expected.begin(), expected.end());
 
@@ -198,7 +194,6 @@ BOOST_AUTO_TEST_CASE(parse_hex)
     // Empty string is supported
     static_assert(""_hex.empty());
     static_assert(""_hex_u8.empty());
-    BOOST_CHECK_EQUAL(""_hex_v.size(), 0);
     BOOST_CHECK_EQUAL(""_hex_v_u8.size(), 0);
     BOOST_CHECK_EQUAL(ParseHex("").size(), 0);
     BOOST_CHECK_EQUAL(TryParseHex<uint8_t>("").value().size(), 0);
diff --git a/src/util/strencodings.h b/src/util/strencodings.h
index f9076de047..2939527a82 100644
--- a/src/util/strencodings.h
+++ b/src/util/strencodings.h
@@ -381,9 +381,6 @@ consteval uint8_t ConstevalHexDigit(const char c)
  * ""_hex is a compile-time user-defined literal returning a
  * `std::array<std::byte>`, equivalent to ParseHex(). Variants provided:
  *
- * - ""_hex_v: Returns `std::vector<std::byte>`, useful for heap allocation or
- *   variable-length serialization.
- *
  * - ""_hex_u8: Returns `std::array<uint8_t>`, for cases where `std::byte` is
  *   incompatible.
  *
@@ -433,9 +430,6 @@ constexpr auto operator""_hex() { return str.bytes; }
 template <util::detail::Hex str>
 constexpr auto operator""_hex_u8() { return std::bit_cast<std::array<uint8_t, str.bytes.size()>>(str.bytes); }
 
-template <util::detail::Hex str>
-constexpr auto operator""_hex_v() { return std::vector<std::byte>{str.bytes.begin(), str.bytes.end()}; }
-
 template <util::detail::Hex str>
 inline auto operator""_hex_v_u8() { return std::vector<uint8_t>{UCharCast(str.bytes.data()), UCharCast(str.bytes.data() + str.bytes.size())}; }
 

```
</details>

I think it might make sense the remove the operator? Since we're dealing with compile-time string literals, I think we shouldn't encourage new users to take vectors instead of arrays?

Reply:
Yup, noticed it earlier as well, but decided to keep for now. Could drop if more people agree.

As noted in the docs, `vector` is currently used to signal that a size-prefix should be added during serialization, so if the conversion to accept `std::byte` in more places happens before adaptation to accept `std::array` for size-prefixed data, `_hex_v` will become necessary.

It also feels more complete to have all `byte`/`uint8_t` and `array`/`vector` combinations.

Reply:
> so if the conversion to accept std::byte in more places happens before adaptation to accept std::array for size-prefixed data, _hex_v will become necessary.

That is a good point, so we are indeed expecting to use it in the future. In that case, since it's already reviewed, let's keep it in and you can mark this as resolved.